File: src/gui/src/renderer/src/model/customRules.ts
/**
 * The custom rules of the rule configuration file, as the editor holds them.
 *
 * The editor has two panes over one value — a form and the raw JSON of the `custom` array — so the
 * text and the structure have to agree exactly: `parse` and `serialise` are inverses over anything
 * the form can produce.
 *
 * Only what is certainly wrong is checked here: the id format, duplicate ids, and the fields the
 * engine requires. A regular expression is never compiled locally, because the engine reads Java
 * syntax and this runtime reads JavaScript syntax; the two disagree on possessive quantifiers,
 * named groups and character-class syntax, and a local verdict would be wrong in both directions.
 * Everything else is the engine's judgement, obtained through `engine:validate-rules`.
 */

import type { CustomMatch, CustomRule } from "../../../shared/rulesFile";

/** `U` followed by 1 to 15 of the characters the engine accepts (docs/rules.md). */
const ID_PATTERN = /^U[0-9A-Za-z_-]{1,15}$/;

/** The severities a custom rule can declare. */
export const CUSTOM_SEVERITIES: readonly string[] = ["HIGH", "MEDIUM", "LOW", "ADVISORY"];

/** The asset kinds a custom rule can examine. */
export const CUSTOM_TARGETS: readonly string[] = ["COBOL", "COPYBOOK", "JCL", "BMS"];

/** The subcommands a custom rule can run under. */
export const CUSTOM_COMMANDS: readonly string[] = ["LINT", "SQL_LINT", "REPORT", "FIX", "SCAN"];

/** How far forward a `checked-after` rule follows the control flow. */
export const CUSTOM_SCOPES: readonly string[] = [
  "untilNextMatchingStatement",
  "untilParagraphEnd",
  "untilProgramEnd",
];

/** What is wrong with one definition. The wording lives in text.ts; this names the problem. */
export type CustomRuleProblemCode =
  | "idFormat"
  | "idDuplicate"
  | "nameRequired"
  | "messageRequired"
  | "regexRequired"
  | "verbRequired"
  | "afterVerbRequired"
  | "dataItemRequired";

export interface CustomRuleProblem {
  readonly code: CustomRuleProblemCode;
  /** Position in the `custom` array, so a definition with no usable id can still be pointed at. */
  readonly index: number;
  readonly id: string;
}

export interface CustomRulesParse {
  /** The definitions, or an empty array when the text could not be read. */
  readonly rules: readonly CustomRule[];
  /**
   * Why the text could not be read; null when it could. It is the JSON parser's own message, or
   * NOT_AN_ARRAY when the text parsed but is not a list of definitions.
   */
  readonly error: string | null;
}

/** The `error` of a text that parses as JSON but is not an array of objects. */
export const NOT_AN_ARRAY = "notAnArray";

/** The default detection of each kind. A new rule and a change of kind both start here. */
export function emptyMatch(kind: CustomMatch["kind"]): CustomMatch {
  switch (kind) {
    case "line":
      return { kind: "line", regex: "", ignoreCase: false, area: "programArea" };
    case "statement":
      return { kind: "statement", verb: [] };
    default:
      return {
        kind: "checked-after",
        after: { verb: "" },
        checks: { dataItem: [] },
        scope: "untilNextMatchingStatement",
        onEveryPath: false,
      };
  }
}

/** A new definition, with an id that does not collide with the ones already written. */
export function emptyCustomRule(existing: readonly CustomRule[]): CustomRule {
  const taken = new Set(existing.map((rule) => rule.id));
  let ordinal = existing.length + 1;
  while (taken.has(`U${String(ordinal).padStart(3, "0")}`)) {
    ordinal += 1;
  }
  return {
    id: `U${String(ordinal).padStart(3, "0")}`,
    name: "",
    message: "",
    severity: "MEDIUM",
    match: emptyMatch("line"),
  };
}

/**
 * The definition's detection. A definition typed by hand can carry no `match` at all, or one of a
 * kind this build does not know; the form still has to render, so it falls back to an empty line
 * match. The original text is not rewritten unless the form is then edited.
 */
export function matchOf(rule: CustomRule): CustomMatch {
  const match = rule.match as Partial<CustomMatch> | undefined;
  if (match === undefined || match === null) {
    return emptyMatch("line");
  }
  switch (match.kind) {
    case "line":
    case "statement":
    case "checked-after":
      return match as CustomMatch;
    default:
      return emptyMatch("line");
  }
}

/** Reads the raw JSON of the `custom` array. */
export function parse(raw: string): CustomRulesParse {
  let value: unknown;
  try {
    value = JSON.parse(raw === "" ? "[]" : raw);
  } catch (error) {
    return { rules: [], error: error instanceof Error ? error.message : String(error) };
  }
  if (!Array.isArray(value)) {
    return { rules: [], error: NOT_AN_ARRAY };
  }
  if (value.some((element) => element === null || typeof element !== "object" || Array.isArray(element))) {
    return { rules: [], error: NOT_AN_ARRAY };
  }
  return { rules: value as CustomRule[], error: null };
}

/** Writes the `custom` array as the text of the raw pane and of the rule file. */
export function serialise(defs: readonly CustomRule[]): string {
  return JSON.stringify(defs.map(pruned), null, 2);
}

/**
 * Drops the optional fields that were left empty.
 *
 * Writing them out would not be harmless: an empty `excludeRegex` matches every line and would
 * silence the rule entirely, and an empty optional string is not always read as "use the default".
 */
export function pruned(rule: CustomRule): CustomRule {
  const result: Record<string, unknown> = { id: rule.id, name: rule.name };
  putText(result, "category", rule.category);
  putText(result, "severity", rule.severity);
  putList(result, "targets", rule.targets);
  putList(result, "commands", rule.commands);
  result["message"] = rule.message;
  putText(result, "summary", rule.summary);
  putText(result, "rationale", rule.rationale);
  putText(result, "remedy", rule.remedy);
  result["match"] = prunedMatch(matchOf(rule));
  return result as unknown as CustomRule;
}

function prunedMatch(match: CustomMatch): CustomMatch {
  if (match.kind === "line") {
    const result: Record<string, unknown> = {
      kind: "line",
      regex: match.regex ?? "",
      ignoreCase: match.ignoreCase === true,
      area: match.area === "wholeLine" ? "wholeLine" : "programArea",
    };
    putText(result, "excludeRegex", match.excludeRegex);
    return result as unknown as CustomMatch;
  }
  if (match.kind === "statement") {
    const result: Record<string, unknown> = { kind: "statement", verb: nonBlank(match.verb) };
    putList(result, "missingClause", match.missingClause);
    putText(result, "inParagraph", match.inParagraph);
    return result as unknown as CustomMatch;
  }
  const after: Record<string, unknown> = { verb: match.after?.verb ?? "" };
  putText(after, "textRegex", match.after?.textRegex);
  return {
    kind: "checked-after",
    after,
    checks: { dataItem: nonBlank(match.checks?.dataItem) },
    scope: match.scope ?? "untilNextMatchingStatement",
    onEveryPath: match.onEveryPath === true,
  } as unknown as CustomMatch;
}

function putText(target: Record<string, unknown>, key: string, value: string | undefined): void {
  if (value !== undefined && value.trim() !== "") {
    target[key] = value;
  }
}

function putList(
  target: Record<string, unknown>,
  key: string,
  value: readonly string[] | undefined,
): void {
  const list = nonBlank(value);
  if (list.length > 0) {
    target[key] = list;
  }
}

function nonBlank(value: readonly string[] | undefined): string[] {
  return (value ?? []).map((element) => element.trim()).filter((element) => element !== "");
}

/**
 * The problems that can be decided without the engine: the id format, duplicate ids, and the fields
 * with no default to fall back on.
 */
export function localProblems(defs: readonly CustomRule[]): CustomRuleProblem[] {
  const problems: CustomRuleProblem[] = [];
  const seen = new Set<string>();
  defs.forEach((rule, index) => {
    const id = rule.id ?? "";
    const at = (code: CustomRuleProblemCode): void => {
      problems.push({ code, index, id });
    };
    if (!ID_PATTERN.test(id)) {
      at("idFormat");
    } else if (seen.has(id)) {
      at("idDuplicate");
    }
    seen.add(id);
    if ((rule.name ?? "").trim() === "") at("nameRequired");
    if ((rule.message ?? "").trim() === "") at("messageRequired");
    const match = matchOf(rule);
    if (match.kind === "line" && (match.regex ?? "").trim() === "") at("regexRequired");
    if (match.kind === "statement" && nonBlank(match.verb).length === 0) at("verbRequired");
    if (match.kind === "checked-after") {
      if ((match.after?.verb ?? "").trim() === "") at("afterVerbRequired");
      if (nonBlank(match.checks?.dataItem).length === 0) at("dataItemRequired");
    }
  });
  return problems;
}

/** The engine's own complaints that name this rule, so they can be shown beside its fields. */
export function engineErrorsFor(errors: readonly string[], id: string): string[] {
  return id === "" ? [] : errors.filter((error) => error.includes(id));
}

File: src/engine/rules/src/main/java/jp/cobolinsight/rules/custom/CustomRules.java
package jp.cobolinsight.rules.custom;

import jp.cobolinsight.core.finding.Severity;
import jp.cobolinsight.core.json.JsonReader;
import jp.cobolinsight.core.rule.Command;
import jp.cobolinsight.core.rule.Needs;
import jp.cobolinsight.core.rule.Rule;
import jp.cobolinsight.core.rule.RuleMeta;
import jp.cobolinsight.core.source.AssetKind;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;
import java.util.stream.Collectors;

/**
 * Builds a {@link Rule} from one {@code custom} entry of {@code rules.json}.
 *
 * <p>Every problem is reported by throwing {@link IllegalArgumentException} with a message the
 * caller shows to the author; the caller drops that one entry and keeps the rest. Messages are
 * Japanese because they are read by the same people who read the rest of the tool.
 *
 * <p>IDs are restricted to {@code U…} so that a custom rule can never collide with a built-in one.
 */
public final class CustomRules {

    /** Custom rule IDs. The U prefix keeps them clear of the built-in R and S ranges. */
    private static final Pattern ID_FORMAT = Pattern.compile("U[0-9A-Za-z_-]{1,15}");

    private static final String DEFAULT_CATEGORY = "利用者定義";

    private static final String DEFAULT_RATIONALE =
            "利用者が定義したルールです。定義に理由が書かれていません。";

    private static final String DEFAULT_REMEDY = "定義した検査の意図に沿って該当箇所を直します。";

    private CustomRules() {
    }

    public static Rule of(Map<String, Object> object) {
        String id = requireString(object, "id");
        if (!ID_FORMAT.matcher(id).matches()) {
            throw new IllegalArgumentException("id は U で始まり、英数字・ハイフン・下線が"
                    + "1〜15文字続く形にする(指定値 " + id + ")");
        }
        String name = requireString(object, "name");
        String message = requireString(object, "message");
        Map<String, Object> match = requireObject(object, "match");
        String kind = requireString(match, "kind");
        Set<AssetKind> targets = toTargets(object.get("targets"));

        return switch (kind) {
            case "line" -> lineRule(object, match, id, name, message, targets);
            case "statement" -> statementRule(object, match, id, name, message, targets);
            case "checked-after" -> checkedAfterRule(object, match, id, name, message, targets);
            default -> throw new IllegalArgumentException("match.kind に扱えない値がある: " + kind
                    + "(扱えるのは line・statement・checked-after)");
        };
    }

    // ---- line ----

    private static Rule lineRule(Map<String, Object> object, Map<String, Object> match, String id,
            String name, String message, Set<AssetKind> targets) {
        String regex = requireString(match, "regex");
        boolean ignoreCase = optionalBoolean(match, "ignoreCase");
        String area = optionalString(match, "area", "programArea");
        if (!area.equals("programArea") && !area.equals("wholeLine")) {
            throw new IllegalArgumentException("match.area に扱えない値がある: " + area
                    + "(扱えるのは programArea・wholeLine)");
        }
        boolean wholeLine = area.equals("wholeLine");
        String excludeText = optionalString(match, "excludeRegex", "");
        Pattern exclude = excludeText.isEmpty() ? null
                : compile(excludeText, ignoreCase, "match.excludeRegex");
        RuleMeta meta = metaOf(object, id, name, targets,
                "正規表現「" + regex + "」に一致する行を検出します。",
                describeLine(targets, regex, ignoreCase, wholeLine, excludeText),
                Set.of(Needs.SOURCE_TEXT));
        return new LineRule(meta, compile(regex, ignoreCase, "match.regex"), exclude, message,
                wholeLine);
    }

    private static String describeLine(Set<AssetKind> targets, String regex, boolean ignoreCase,
            boolean wholeLine, String exclude) {
        StringBuilder out = new StringBuilder();
        out.append("対象は ").append(labelOf(targets))
                .append(" の各行です。正規表現「").append(regex)
                .append("」に一致する行を検出します(")
                .append(ignoreCase ? "大小を区別しません" : "大小を区別します")
                .append(")。");
        if (wholeLine) {
            out.append("行全体を対象とし、注記行も走査します。");
        } else {
            out.append("COBOL 本体とコピー句では注記行を除き、8〜72桁の範囲を対象とします。");
        }
        if (!exclude.isEmpty()) {
            out.append("同じ行が正規表現「").append(exclude).append("」にも一致する場合は除きます。");
        }
        return out.toString();
    }

    // ---- statement ----

    private static Rule statementRule(Map<String, Object> object, Map<String, Object> match,
            String id, String name, String message, Set<AssetKind> targets) {
        List<String> verbs = requireStrings(match, "verb").stream()
                .map(verb -> verb.toUpperCase(Locale.ROOT)).toList();
        List<String> clauses = optionalStrings(match, "missingClause").stream()
                .map(clause -> clause.toUpperCase(Locale.ROOT)).toList();
        String inParagraphText = optionalString(match, "inParagraph", "");
        Pattern inParagraph = inParagraphText.isEmpty() ? null
                : compile(inParagraphText, true, "match.inParagraph");
        RuleMeta meta = metaOf(object, id, name, targets,
                String.join("・", verbs) + " 文を検出します。",
                describeStatement(verbs, clauses, inParagraphText), Set.of(Needs.SEMANTIC));
        return new StatementRule(meta, verbs, clauses, inParagraph, message);
    }

    private static String describeStatement(List<String> verbs, List<String> clauses,
            String inParagraph) {
        StringBuilder out = new StringBuilder("対象は ").append(String.join("・", verbs))
                .append(" 文です。");
        if (clauses.isEmpty()) {
            out.append("該当する文をすべて検出します。");
        } else {
            out.append(String.join("・", clauses)).append(" のいずれの句も伴わないものを検出します。");
        }
        if (!inParagraph.isEmpty()) {
            out.append("段落名が正規表現「").append(inParagraph).append("」に一致する段落だけを見ます。");
        }
        return out.toString();
    }

    // ---- checked-after ----

    private static Rule checkedAfterRule(Map<String, Object> object, Map<String, Object> match,
            String id, String name, String message, Set<AssetKind> targets) {
        Map<String, Object> after = requireObject(match, "after");
        String verb = requireString(after, "verb");
        String textRegex = optionalString(after, "textRegex", "");
        Pattern afterText = textRegex.isEmpty() ? null
                : compile(textRegex, false, "match.after.textRegex");
        Map<String, Object> checks = requireObject(match, "checks");
        List<String> dataItems = toDataItems(checks.get("dataItem"));
        CheckedAfterRule.Scope scope = toScope(optionalString(match, "scope",
                "untilNextMatchingStatement"));
        boolean onEveryPath = optionalBoolean(match, "onEveryPath");
        RuleMeta meta = metaOf(object, id, name, targets,
                verb + " の実行後に " + String.join("・", dataItems) + " を検査しない箇所を検出します。",
                describeCheckedAfter(verb, textRegex, dataItems, scope, onEveryPath),
                Set.of(Needs.SEMANTIC, Needs.CFG));
        return new CheckedAfterRule(meta, verb, afterText, dataItems, scope, onEveryPath, message);
    }

    private static String describeCheckedAfter(String verb, String textRegex,
            List<String> dataItems, CheckedAfterRule.Scope scope, boolean onEveryPath) {
        StringBuilder out = new StringBuilder("対象は ").append(verb).append(" 文");
        if (!textRegex.isEmpty()) {
            out.append("(本文が正規表現「").append(textRegex).append("」に一致するもの)");
        }
        out.append("です。その実行後、").append(switch (scope) {
            case UNTIL_NEXT_MATCHING_STATEMENT -> "次の同じ動詞の文に達するまで";
            case UNTIL_PARAGRAPH_END -> "段落の終わりまで";
            case UNTIL_PROGRAM_END -> "プログラムの終わりまで";
        }).append("の前方経路で ").append(String.join("・", dataItems))
                .append(" を条件参照しないものを検出します(")
                .append(onEveryPath ? "すべての経路で検査を要する" : "いずれかの経路に検査があれば足りる")
                .append(")。");
        return out.toString();
    }

    private static CheckedAfterRule.Scope toScope(String text) {
        return switch (text) {
            case "untilNextMatchingStatement" ->
                    CheckedAfterRule.Scope.UNTIL_NEXT_MATCHING_STATEMENT;
            case "untilParagraphEnd" -> CheckedAfterRule.Scope.UNTIL_PARAGRAPH_END;
            case "untilProgramEnd" -> CheckedAfterRule.Scope.UNTIL_PROGRAM_END;
            default -> throw new IllegalArgumentException("match.scope に扱えない値がある: " + text
                    + "(扱えるのは untilNextMatchingStatement・untilParagraphEnd・untilProgramEnd)");
        };
    }

    private static List<String> toDataItems(Object value) {
        List<String> items = new ArrayList<>();
        if (value instanceof String single) {
            items.add(single.strip());
        } else if (value != null) {
            for (Object element : JsonReader.asArray(value)) {
                if (!(element instanceof String text) || text.isBlank()) {
                    throw new IllegalArgumentException("checks.dataItem は空でない文字列で書く");
                }
                items.add(text.strip());
            }
        }
        if (items.isEmpty()) {
            throw new IllegalArgumentException("checks.dataItem が無い、または空である");
        }
        return items;
    }

    // ---- shared ----

    private static RuleMeta metaOf(Map<String, Object> object, String id, String name,
            Set<AssetKind> targets, String defaultSummary, String detection, Set<Needs> needs) {
        return RuleMeta.named(id, name, optionalString(object, "category", DEFAULT_CATEGORY))
                .summary(optionalString(object, "summary", defaultSummary))
                .rationale(optionalString(object, "rationale", DEFAULT_RATIONALE))
                .detection(detection)
                .remedy(optionalString(object, "remedy", DEFAULT_REMEDY))
                .example(optionalString(object, "badExample", ""),
                        optionalString(object, "goodExample", ""))
                .severity(toSeverity(optionalString(object, "severity", "MEDIUM")))
                .commands(toCommands(object.get("commands")))
                .targets(targets)
                .needs(needs)
                .build();
    }

    private static Severity toSeverity(String text) {
        try {
            return Severity.valueOf(text.toUpperCase(Locale.ROOT));
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("severity に扱えない値がある: " + text
                    + "(扱えるのは HIGH・MEDIUM・LOW・ADVISORY)");
        }
    }

    /** Absent means lint and report, which is where a hand-written rule was run before V2. */
    private static Set<Command> toCommands(Object value) {
        if (value == null) {
            return Set.of(Command.LINT, Command.REPORT);
        }
        EnumSet<Command> commands = EnumSet.noneOf(Command.class);
        for (Object element : asArray(value, "commands")) {
            if (!(element instanceof String text)) {
                throw new IllegalArgumentException("commands に扱えない値がある: " + element);
            }
            try {
                commands.add(Command.valueOf(text.strip().toUpperCase(Locale.ROOT)));
            } catch (IllegalArgumentException e) {
                throw new IllegalArgumentException("commands に扱えない値がある: " + text
                        + "(扱えるのは LINT・SQL_LINT・REPORT・FIX・SCAN)");
            }
        }
        if (commands.isEmpty()) {
            throw new IllegalArgumentException("commands が空である");
        }
        return commands;
    }

    private static Set<AssetKind> toTargets(Object value) {
        if (value == null) {
            return Set.of(AssetKind.COBOL);
        }
        EnumSet<AssetKind> targets = EnumSet.noneOf(AssetKind.class);
        for (Object element : asArray(value, "targets")) {
            if (!(element instanceof String text)) {
                throw new IllegalArgumentException("targets に扱えない種別がある: " + element);
            }
            try {
                targets.add(AssetKind.valueOf(text.strip().toUpperCase(Locale.ROOT)));
            } catch (IllegalArgumentException e) {
                throw new IllegalArgumentException("targets に扱えない種別がある: " + text
                        + "(扱えるのは COBOL・COPYBOOK・BMS・JCL)");
            }
        }
        if (targets.isEmpty()) {
            throw new IllegalArgumentException("targets が空である");
        }
        return targets;
    }

    private static String labelOf(Set<AssetKind> targets) {
        return targets.stream().map(Enum::name).sorted().collect(Collectors.joining("・"));
    }

    private static Pattern compile(String text, boolean ignoreCase, String field) {
        try {
            return Pattern.compile(text, ignoreCase ? Pattern.CASE_INSENSITIVE : 0);
        } catch (PatternSyntaxException e) {
            throw new IllegalArgumentException(
                    field + " の正規表現を解釈できない: " + e.getDescription());
        }
    }

    private static List<Object> asArray(Object value, String field) {
        try {
            return JsonReader.asArray(value);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(field + " は配列で書く");
        }
    }

    private static Map<String, Object> requireObject(Map<String, Object> object, String field) {
        Object value = object.get(field);
        if (value == null) {
            throw new IllegalArgumentException(field + " が無い");
        }
        try {
            return JsonReader.asObject(value);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException(field + " はオブジェクトで書く");
        }
    }

    private static String requireString(Map<String, Object> object, String field) {
        Object value = object.get(field);
        if (!(value instanceof String text) || text.isBlank()) {
            throw new IllegalArgumentException(field + " が無い、または空である");
        }
        return text.strip();
    }

    private static List<String> requireStrings(Map<String, Object> object, String field) {
        List<String> values = optionalStrings(object, field);
        if (values.isEmpty()) {
            throw new IllegalArgumentException(field + " が無い、または空である");
        }
        return values;
    }

    private static List<String> optionalStrings(Map<String, Object> object, String field) {
        Object value = object.get(field);
        if (value == null) {
            return List.of();
        }
        if (value instanceof String single) {
            return single.isBlank() ? List.of() : List.of(single.strip());
        }
        List<String> values = new ArrayList<>();
        for (Object element : asArray(value, field)) {
            if (!(element instanceof String text) || text.isBlank()) {
                throw new IllegalArgumentException(field + " は空でない文字列の配列で書く");
            }
            values.add(text.strip());
        }
        return values;
    }

    private static String optionalString(Map<String, Object> object, String field,
            String fallback) {
        Object value = object.get(field);
        if (value == null) {
            return fallback;
        }
        if (!(value instanceof String text)) {
            throw new IllegalArgumentException(field + " は文字列で書く");
        }
        return text.isBlank() ? fallback : text.strip();
    }

    private static boolean optionalBoolean(Map<String, Object> object, String field) {
        Object value = object.get(field);
        if (value == null) {
            return false;
        }
        if (!(value instanceof Boolean flag)) {
            throw new IllegalArgumentException(field + " は true か false で書く");
        }
        return flag;
    }
}
